### PR TITLE
Add Microsoft Store REST Source as default option and fix telemetry gaps

### PR DESCRIPTION
**Pull request by sachintaMSFT**
Add Microsoft Store REST Source as default option
Fix gaps found in telemetry from data validation

- [ ] Have you signed the [Contributor License Agreement](https://cla.opensource.microsoft.com/microsoft/winget-pkgs)?
- [ ] Are you working against an Issue?

-----

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/microsoft/winget-cli/pull/1216)

### CODE REVIEW

File: src/AppInstallerCommonCore/AppInstallerTelemetry.cpp [hunk 1]
```diff
@@ -179,6 +179,8 @@ namespace AppInstaller::Logging
             packageVersion = Runtime::GetPackageVersion();
         }
 
+        Logging::SetActivityId();
```
Comment (JohnMcPMS) [inline src/AppInstallerCommonCore/AppInstallerTelemetry.cpp:182]: Why is this moved here?

File: src/AppInstallerRepositoryCore/RepositorySource.cpp [hunk 2]
```diff
@@ -417,6 +426,16 @@ namespace AppInstaller::Repository
                     details.TrustLevel = SourceTrustLevel::Trusted | SourceTrustLevel::StoreOrigin;
                     result.emplace_back(std::move(details));
                 }
+
+                if (IsMSStoreDefaultSourceEnabled())
```
Comment (JohnMcPMS) [inline src/AppInstallerRepositoryCore/RepositorySource.cpp:430]: I don't want this included in the default list if it doesn't support search.  It also needs to be defined in the appropriate "get details" rather than here.

The easiest way to make it work all up would be a new bool on the details that allows it to be marked as hidden (maybe a better word to indicate that it requires direct targeting), then filter that out when creating the default composite source and returning details.  That lets it be targeted directly without making it part of the default set, as well as being able to update the details if needed.

@denelon, what would be a good word to describe a source that is not included with `winget search foo`, but can still be accessed via `winget search foo -s source`?  Hidden?  Shadow?  Explicit?

File: src/AppInstallerRepositoryCore/RepositorySource.cpp [hunk 3]
```diff
@@ -417,6 +426,16 @@ namespace AppInstaller::Repository
...
                     result.emplace_back(std::move(details));
                 }
+
+                if (IsMSStoreDefaultSourceEnabled())
+                {
+                    SourceDetailsInternal details;
+                    details.Name = s_Source_MSStoreDefault_Name;
+                    details.Type = Rest::RestSourceFactory::Type();
+                    details.Arg = s_Source_MSStoreDefault_Arg;
+                    details.TrustLevel = SourceTrustLevel::Trusted | SourceTrustLevel::StoreOrigin;
```
Comment (JohnMcPMS) [inline src/AppInstallerRepositoryCore/RepositorySource.cpp:436]: `StoreOrigin` is only useful for packaged index sources.

File: src/AppInstallerRepositoryCore/RepositorySource.cpp [hunk 4]
```diff
@@ -417,6 +426,16 @@ namespace AppInstaller::Repository
                     details.TrustLevel = SourceTrustLevel::Trusted | SourceTrustLevel::StoreOrigin;
                     result.emplace_back(std::move(details));
                 }
+
+                if (IsMSStoreDefaultSourceEnabled())
+                {
+                    SourceDetailsInternal details;
```
Comment (JohnMcPMS) [inline src/AppInstallerRepositoryCore/RepositorySource.cpp:432]: Need to set the `Id`.

**Review by JohnMcPMS — CHANGES_REQUESTED** [review]

File: src/AppInstallerCommonCore/AppInstallerTelemetry.cpp (hunk 1 above)
Comment (sachintaMSFT) [reply src/AppInstallerCommonCore/AppInstallerTelemetry.cpp:182]: Earlier commit missed setting activity Id for winget-cli. I did that in our preview package, though. So, commiting it here and Instead of setting ActivityID at two different spots for COM and cli paths, I felt it's better to do at one spot, here.

**Comment by sreadingMSFT** [comment]
> 

Probably might as well pull in the com interface changes as well in this checkin.

---
Refers to: src/AppInstallerRepositoryCore/RepositorySource.cpp:1098 in 5e6c3af. [](commit_id = 5e6c3afb0962446db1e9ac0ef476150a6f8aab77, deletion_comment = False)

File: src/AppInstallerCommonCore/AppInstallerTelemetry.cpp (hunk 1 above)
Comment (JohnMcPMS) [reply src/AppInstallerCommonCore/AppInstallerTelemetry.cpp:182]: It is probably fine for now; we should just make it part of the object construction in the "make not global" PR.

File: src/AppInstallerRepositoryCore/RepositorySource.cpp (hunk 2 above)
Comment (JohnMcPMS) [reply src/AppInstallerRepositoryCore/RepositorySource.cpp:430]: I think I like `Restricted` the most from @denelon 's suggestions.  This obviously isn't permanent anyway, but let's use it as the starting point.

File: src/AppInstallerCommonCore/AppInstallerTelemetry.cpp (hunk 1 above)
Comment (sachintaMSFT) [reply src/AppInstallerCommonCore/AppInstallerTelemetry.cpp:182]: Sure. That is the plan with that pending code change.

File: src/AppInstallerRepositoryCore/RepositorySource.cpp [hunk 5]
```diff
@@ -417,6 +426,17 @@ namespace AppInstaller::Repository
...
+
+                if (IsMSStoreDefaultSourceEnabled())
+                {
+                    SourceDetailsInternal details;
+                    details.Name = s_Source_MSStoreDefault_Name;
+                    details.Type = Rest::RestSourceFactory::Type();
+                    details.Arg = s_Source_MSStoreDefault_Arg;
+                    details.Identifier = s_Source_MSStoreDefault_Identifier;
+                    details.TrustLevel = SourceTrustLevel::Trusted;
+                    result.emplace_back(std::move(details));
```
Comment (JohnMcPMS) [inline src/AppInstallerRepositoryCore/RepositorySource.cpp:438]: This should all come from just calling `GetWellKnownSourceDetails` rather than being implemented twice.

**Review by JohnMcPMS — CHANGES_REQUESTED** [review]
We also need to make the COM API work by default on Win11 insider builds to continue to enable seemless integration.  I would suggest just making the experimental feature return true if the OS version is high enough (there is a function to check that in Runtime).

File: src/AppInstallerRepositoryCore/RepositorySource.cpp [hunk 6]
```diff
@@ -1073,6 +1094,17 @@ namespace AppInstaller::Repository
...
             return details;
         }
+        case WellKnownSource::MicrosoftStore:
+        {
+            SourceDetailsInternal details;
+            details.Name = s_Source_MSStoreDefault_Name;
+            details.Type = Rest::RestSourceFactory::Type();
+            details.Arg = s_Source_MSStoreDefault_Arg;
+            details.Identifier = s_Source_MSStoreDefault_Identifier;
+            details.TrustLevel = SourceTrustLevel::Trusted | SourceTrustLevel::StoreOrigin;
```
Comment (sreadingMSFT) [inline src/AppInstallerRepositoryCore/RepositorySource.cpp:1104]: > | SourceTrustLevel::StoreOrigin [](http://example.com/codeflow?start=58&length=32)

Should remove it here too.

File: src/AppInstallerRepositoryCore/RepositorySource.cpp [hunk 7]
```diff
@@ -1073,6 +1089,18 @@ namespace AppInstaller::Repository
             details.TrustLevel = SourceTrustLevel::Trusted | SourceTrustLevel::StoreOrigin;
             return details;
         }
+        case WellKnownSource::MicrosoftStore:
+        {
+            SourceDetailsInternal details;
+            details.Name = s_Source_MSStoreDefault_Name;
```
Comment (JohnMcPMS) [inline src/AppInstallerRepositoryCore/RepositorySource.cpp:1095]: Should have `Origin` set.

File: src/AppInstallerRepositoryCore/RepositorySource.cpp [hunk 8]
```diff
@@ -84,6 +92,11 @@ namespace AppInstaller::Repository
             return IsDefaultSourceEnabled(s_Source_WingetMSStoreDefault_Name, ExperimentalFeature::Feature::ExperimentalMSStore, onlyExplicit, TogglePolicy::Policy::MSStoreSource);
         }
 
+        bool IsMSStoreDefaultSourceEnabled(bool onlyExplicit = false)
+        {
+            return IsDefaultSourceEnabled(s_Source_MSStoreDefault_Name, ExperimentalFeature::Feature::None, onlyExplicit, TogglePolicy::Policy::MSStoreSource);
```
Comment (JohnMcPMS) [inline src/AppInstallerRepositoryCore/RepositorySource.cpp:97]:
```suggestion
            return IsDefaultSourceEnabled(s_Source_MSStoreDefault_Name, ExperimentalFeature::Feature::ExperimentalMSStore, onlyExplicit, TogglePolicy::Policy::MSStoreSource);

```

**Review by JohnMcPMS — CHANGES_REQUESTED** [review]
Per discussion, make the source part of the store EF.

The problem with this is it more clearly shows the issue with `OpenSourceFromDetails`.  That function needs to go away, but a larger redesign of the interaction is needed.

You will need to make `OpenSourceFromDetails` handle not finding the source that it is looking for.  Right now I would suggest that be by simply skipping the update step.

File: src/AppInstallerRepositoryCore/RepositorySource.cpp [hunk 9]
```diff
@@ -926,6 +937,11 @@ namespace AppInstaller::Repository
                 bool sourceUpdated = false;
                 for (auto& source : currentSources)
                 {
+                    if (source.get().Restricted == true)
```
Comment (sreadingMSFT) [inline src/AppInstallerRepositoryCore/RepositorySource.cpp:940]: > if (source.get().Restricted == true) [](http://example.com/codeflow?start=19&length=37)

Can you add a comment here like:
// Restricted sources may not support the full set of functionality so they shouldn't be included in the default aggregated source.

**Review by sreadingMSFT — APPROVED** [review]
:shipit:

File: src/AppInstallerRepositoryCore/RepositorySource.cpp [hunk 10]
```diff
@@ -1026,7 +1039,8 @@ namespace AppInstaller::Repository
                 AICLI_LOG(Repo, Info, << "Named source no longer found. Source was tombstoned: " << details.Name);
                 return {};
             }
-            if (ShouldUpdateBeforeOpen(*source))
+            if (source->Restricted != true &&
```
Comment (JohnMcPMS) [inline src/AppInstallerRepositoryCore/RepositorySource.cpp:1042]: nit: binary operators on booleans are more difficult to understand (for me at least)
```
source->Restricted != true
```
vs
```
!source->Restricted
```

File: src/AppInstallerRepositoryCore/RepositorySource.cpp [hunk 11]
```diff
@@ -1026,7 +1039,8 @@ namespace AppInstaller::Repository
                 AICLI_LOG(Repo, Info, << "Named source no longer found. Source was tombstoned: " << details.Name);
```
Comment (JohnMcPMS) [inline src/AppInstallerRepositoryCore/RepositorySource.cpp:1039]: I know this isn't part of your change, but I now realize that this results in the ability to remove the store source despite the fact that we want the well known sources to *always* be present.  I would suggest removing the Tombstone block.

**Review by JohnMcPMS — APPROVED** [review]

**Comment by mohammedpanamai** [comment]
> Add Microsoft Store REST Source as default option Fix gaps found in telemetry from data validation
> 
> * [ ]  Have you signed the [Contributor License Agreement](https://cla.opensource.microsoft.com/microsoft/winget-pkgs)?
> * [ ]  Are you working against an Issue?
> 
> ###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/microsoft/winget-cli/pull/1216)
